include/crocoddyl/core/integr-action-base.hpp

```diff
@@ -0,0 +1,105 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2021, LAAS-CNRS, University of Edinburgh, University of Oxford,
+//                     University of Trento
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef CROCODDYL_CORE_INTEGRATED_ACTION_BASE_HPP_
+#define CROCODDYL_CORE_INTEGRATED_ACTION_BASE_HPP_
+
+#include "crocoddyl/core/fwd.hpp"
+#include "crocoddyl/core/action-base.hpp"
+#include "crocoddyl/core/diff-action-base.hpp"
+#include "crocoddyl/core/control-base.hpp"
+
+namespace crocoddyl {
+
+template <typename _Scalar>
+class IntegratedActionModelAbstractTpl : public ActionModelAbstractTpl<_Scalar> {
+ public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef _Scalar Scalar;
+  typedef MathBaseTpl<Scalar> MathBase;
+  typedef ActionModelAbstractTpl<Scalar> Base;
+  typedef IntegratedActionDataEulerTpl<Scalar> Data;
+  typedef ActionDataAbstractTpl<Scalar> ActionDataAbstract;
+  typedef DifferentialActionModelAbstractTpl<Scalar> DifferentialActionModelAbstract;
+  typedef ControlAbstractTpl<Scalar> ControlAbstract;
+  typedef typename MathBase::VectorXs VectorXs;
+  typedef typename MathBase::MatrixXs MatrixXs;
+
+  IntegratedActionModelAbstractTpl(boost::shared_ptr<DifferentialActionModelAbstract> model,
+                                  const Scalar time_step = Scalar(1e-3), const bool with_cost_residual = true);
+  IntegratedActionModelAbstractTpl(boost::shared_ptr<DifferentialActionModelAbstract> model,
+                                  boost::shared_ptr<ControlAbstract> control,
+                                  const Scalar time_step = Scalar(1e-3), const bool with_cost_residual = true);
+  virtual ~IntegratedActionModelAbstractTpl();
+
+  const boost::shared_ptr<DifferentialActionModelAbstract>& get_differential() const;
+  const Scalar get_dt() const;
+
+  void set_dt(const Scalar dt);
+  void set_differential(boost::shared_ptr<DifferentialActionModelAbstract> model);
+
+  /**
+   * @brief Return the dimension of the control input of the differential action model
+   */
+  std::size_t get_nu_diff() const;
+
+ protected:
+  using Base::has_control_limits_;  //!< Indicates whether any of the control limits are active
+  using Base::nu_;                  //!< Dimension of the control
+  using Base::nr_;                  //!< Dimension of the cost residual
+  using Base::state_;               //!< Model of the state
+  using Base::u_lb_;                //!< Lower control limits
+  using Base::u_ub_;                //!< Upper control limits
+  using Base::unone_;               //!< Neutral state
+
+  void init();
```

Comment:
The reason is to avoid duplicating the code that initializes the object in the two constructors of the class. Actually I've seen this done many times and I don't see any downside.

Reply:
I understand the idea behind `init`, and we can keep this approach. However, you're still a bit duplicating the code as `IntegratedActionModelEulerTpl::init` implements the same code than the `IntegratedActionModelAbstractTpl::init()`.

 Note that when you construct your derived integrated-action, you call `Base(...)` which in turn calls `IntegratedActionModelAbstractTpl::init()`.

Similarly, in the `IntegratedActionModelRK4Tpl::init()`, there are bits that have been already implemented in the `IntegratedActionModelAbstractTpl::init()`. These lines of code needs to be removed: https://github.com/andreadelprete/crocoddyl/blob/polynomial-control/include/crocoddyl/core/integrator/rk4.hxx#L34-L45